core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js

```diff
@@ -121,6 +123,58 @@ oppia.directive('schemaBasedUnicodeEditor', [
             return $sce.trustAsHtml(
               $filter('convertUnicodeWithParamsToHtml')($scope.localValue));
           };
+
+
+          $scope.allowSpeechRecognition = (
+            $scope.uiConfig() &&
+            !!$window.webkitSpeechRecognition);
+
+          $scope.isCurrentlyRecording = false;
+
+          var recognition = null;
+
+          if ($scope.allowSpeechRecognition) {
+            recognition = new webkitSpeechRecognition();
+          }
+
+          $scope.toggleSpeechRecognition = function() {
+            if ($scope.isCurrentlyRecording) {
+              recognition.stop();
+              $scope.isCurrentlyRecording = false;
+              return;
+            }
+
+            recognition = new webkitSpeechRecognition();
+            recognition.continuous = false;
+            recognition.interimResults = true;
+            var explorationLanguageCode = (
+                    ExplorationPlayerService.getExplorationLanguageCode());
+            recognition.lang = explorationLanguageCode;
```

Comment:
What happens if speech recognition is not supported for the exploration's language code?

Reply:
Should I add a default language English for such cases?

Reply:
In that case, I would say the option to use speech recognition input simply should not be available

Reply:
Thanks for the suggestion! I will make the changes!

Reply:
Hi @tjiang11. I have checked for the language support in webkit speech recognition. It supports all the languages we currently have. Also, there is no specific way to obtain a list of supported languages. In case the language is not supported it automatically switches to English. So, shall I create a separate list of supported languages for checking this? Can you please provide some suggestions on this!
Here is a link for the same:https://stackoverflow.com/questions/14257598/what-are-language-codes-in-chromes-implementation-of-the-html5-speech-recogniti

Reply:
Yes, I think it'd be best if we explicitly define the languages we support for speech recognition. I think it'd be sufficient to just do English for now, but feel free to do more. We don't want it defaulting because if we do end up having explorations in languages that aren't supported, such as perhaps a purely Hinglish exploration, then we don't want it to be taking in input in English.

Reply:
Is it possible to detect whether a specific language is supported?

Reply:
I searched for this and I did not find an explicit way to check if a language is supported. 

Also, the languages we currently have as options in an exploration's language are all supported by webkit speech recognition. 

So, we can also avoid checking this. But if we want to check this, I feel we need an explicit list of supported language codes and then check if the exploration's language code is present in that list.

Reply:
Okay, yes I think that would be fine (having our own list of supported languages).

Reply:
So, just to confirm once:

- We have to store a list of webkit supported languages
- If the code for exploration language is found in the list, we provide the option for speech recognition otherwise don't.

Reply:
The list of the languages would be quite long though.

Reply:
I think that's okay, you can see in constants.js there is quite a long list of languages.

Reply:
Also, should I add a list of only language codes or codes with their description since for checking only the codes would be needed?

Reply:
Whichever is necessary to perform the check, so I think just the language codes is fine